Simplify selectors

Databean opened the pull request:
Pushes down some hardcoded constants and replaces some object-oriented interfaces with imperative interfaces.

Bug: b/389804379
Test: /usr/bin/bazel run :cvd -- --instance_name=1 status

jemoreira commented on base/cvd/cuttlefish/host/commands/cvd/cli/frontline_parser.h:
```diff
@@ -39,46 +32,7 @@ Result<cvd_common::Args> ExtractCvdArgs(cvd_common::Args& args);
-  const std::vector<std::string> internal_cmds_;
-  std::unique_ptr<ArgumentsSeparator> arguments_separator_;
-};
+Result<cvd_common::Args> ExtractCvdArgs(cvd_common::Args& args);
```
Please update the comment over this to explicitly mention that it returns the cvd arguments after removing them from the input parameter.
That comment seems to have been originally for the FrontlineParser class, which did more than this function.

jemoreira commented on base/cvd/cuttlefish/host/commands/cvd/cli/frontline_parser.cpp:
```diff
@@ -17,119 +17,40 @@
-const cvd_common::Args& FrontlineParser::SubCmdArgs() const {
-  return arguments_separator_->SubCmdArgs();
-}
+  cvd_common::Args cmd_args{separated_arguments.sub_cmd_args};
```
Why copy the arguments? can't `cmd_args` be a reference?

Databean replied:
Made it a `const` reference.

jemoreira commented on base/cvd/cuttlefish/host/commands/cvd/cli/frontline_parser.cpp:
```diff
@@ -17,119 +17,40 @@
+  cvd_common::Args new_exec_args{separated_arguments.prog_path};
 
-  cvd_common::Args new_exec_args{prog_path};
+  const auto new_sub_cmd = separated_arguments.sub_cmd;
```
nit: `s/new_sub_cmd/sub_cmd`. I wasted some time searching for the `old_sub_cmd` here.

Databean replied:
Done.

jemoreira commented on base/cvd/cuttlefish/host/commands/cvd/cli/frontline_parser.cpp:
```diff
@@ -17,119 +17,40 @@
-const cvd_common::Args& FrontlineParser::SubCmdArgs() const {
-  return arguments_separator_->SubCmdArgs();
-}
+  cvd_common::Args cmd_args{separated_arguments.sub_cmd_args};
```
nit: `s/cmd_args/sub_cmd_args/`

Databean replied:
Done.

jemoreira commented on base/cvd/cuttlefish/host/commands/cvd/cli/selector/arguments_lexer.cpp:
```diff
@@ -33,6 +34,86 @@ bool Included(const std::string& item, Sets&&... containers) {
+  using FlagPatterns = ArgumentsLexer::FlagPatterns;
+
+ public:
+  static Result<std::unique_ptr<ArgumentsLexer>> Build();
```
Why can't this be a static function of `ArgumentsLexer` itself? Probably renamed as `Create`.

Databean replied:
Done. It's still called `Build`.

jemoreira commented on base/cvd/cuttlefish/host/commands/cvd/cli/selector/arguments_lexer.cpp:
```diff
@@ -79,8 +160,12 @@ ArgumentsLexerBuilder::GenerateFlagPatterns(
-Result<std::unique_ptr<ArgumentsLexer>> ArgumentsLexerBuilder::Build(
-    const LexerFlagsSpecification& known_flags) {
+Result<std::unique_ptr<ArgumentsLexer>> ArgumentsLexerBuilder::Build() {
+  LexerFlagsSpecification known_flags{
+      .known_boolean_flags = {SelectorFlags::kVerbosity},
+      .known_value_flags = {SelectorFlags::kGroupName,
+                            SelectorFlags::kInstanceName},
```
Probably not for this PR, but knowledge of what these flags are meant to be is duplicated between here and `ParseCommonSelectorArguments` in `selector_common_parser.cpp`. One way to fix that would be for this parser to return the cvd flags already parsed instead of a list of strings that need to be parsed again.

Databean replied:
Added a comment in both locations referencing the other place to update.

This is a downside of pushing down hardcoded values. This might be possible to address later in deduplicating `CvdFlag` and `Flag.

jemoreira commented on base/cvd/cuttlefish/host/commands/cvd/cli/selector/arguments_lexer.cpp:
```diff
@@ -33,6 +34,86 @@ bool Included(const std::string& item, Sets&&... containers) {
+  using FlagPatterns = ArgumentsLexer::FlagPatterns;
+
+ public:
+  static Result<std::unique_ptr<ArgumentsLexer>> Build();
```
Why return a `Result`? There is no configuration or randomization for this builder, if it builds once it should always build, right?
My guess is that the `CF_EXPECT`s inside this function should probably be `CHECK` since those would indicate errors to be catch on the first run.

Databean replied:
I don't see the harm in using `Result`s throughout considering that the public interface `TokenizeArguments` can fail based on user input anyway.

In general, I prefer to leave it up to the caller to decide whether to abort, whatever the condition is.

jemoreira commented on base/cvd/cuttlefish/host/commands/cvd/cli/selector/arguments_lexer.cpp:
```diff
@@ -206,5 +291,16 @@ Result<std::vector<std::string>> ArgumentsLexer::Preprocess(
+  ArgumentsLexerBuilder builder;
+
+  std::unique_ptr<ArgumentsLexer> lexer = CF_EXPECT(builder.Build());
+  CF_EXPECT(lexer.get());
```
`Build()` should return `Result<ArgumentsLexer>` instead of the pointer, so this line wouldn't be needed.

Databean replied:
Done.

jemoreira commented on base/cvd/cuttlefish/host/commands/cvd/cli/selector/arguments_lexer.cpp:
```diff
@@ -206,5 +291,16 @@ Result<std::vector<std::string>> ArgumentsLexer::Preprocess(
+  CF_EXPECT(lexer.get());
+
+  std::vector<ArgToken> as_vector = CF_EXPECT(lexer->Tokenize(args));
+  return std::deque<ArgToken>(as_vector.begin(), as_vector.end());
```
nit: I actually prefer this function to return a `vector` instead of a `deque`. There is no difference performance-wise, it's just that `deque` is used for a very specific purpose (efficient insertion and deletion at the front) which is not needed here. I prefer to introduce the `deque` at the caller site where the need for it is evident.

Databean replied:
Done.

Databean commented:
Additional changes in the most recent push not related to comments.

- Corrects division of flags: there are actually three "value" flags, and no boolean flags.
- Deleted parsing support for boolean flags.
- Fixed tests.

jemoreira commented on base/cvd/cuttlefish/host/commands/cvd/cli/frontline_parser.cpp:
```diff
@@ -17,119 +17,53 @@
-  arguments_separator_ = CF_EXPECT(CallSeparator());
-  return {};
-}
+/* the very first command line parser
+ *
+ * Being aware of valid subcommands and cvd-specific commands, it will
+ * separate the command line arguments into:
+ *
+ *  1. program path/name
+ *  2. cvd-specific arguments
+ *     a) selector flags
+ *     b) non-selector flags
+ *  3. subcommand
+ *  4. subcommand arguments
+ */
```
I think this applies to `SeparateArguments` instead.

Databean replied:
Fixed.

Databean commented:
Thanks for reviewing!
